system/lib/pthread/proxying.c

```diff
@@ -0,0 +1,329 @@
+  task_queue* task_queues;
+  int size;
+  int capacity;
+};
+
+static em_proxying_queue system_proxying_queue = {.mutex =
+                                                    PTHREAD_MUTEX_INITIALIZER,
+                                                  .task_queues = NULL,
+                                                  .size = 0,
+                                                  .capacity = 0};
+
+em_proxying_queue* emscripten_proxy_get_system_queue(void) {
+  return &system_proxying_queue;
+}
+
+em_proxying_queue* em_proxying_queue_create(void) {
+  em_proxying_queue* q = malloc(sizeof(em_proxying_queue));
+  if (q == NULL) {
+    return NULL;
+  }
+  *q = (em_proxying_queue){.mutex = PTHREAD_MUTEX_INITIALIZER,
+                           .task_queues = NULL,
+                           .size = 0,
+                           .capacity = 0};
+  return q;
+}
+
+void em_proxying_queue_destroy(em_proxying_queue* q) {
+  assert(q != NULL);
+  assert(q != &system_proxying_queue && "cannot destroy system proxying queue");
+  // No need to acquire the lock; no one should be racing with the destruction
+  // of the queue.
+  pthread_mutex_destroy(&q->mutex);
+  for (int i = 0; i < q->size; i++) {
+    task_queue_deinit(&q->task_queues[i]);
+  }
+  free(q->task_queues);
+  free(q);
+}
+
+// Not thread safe. Returns -1 if there are no tasks for the thread.
+static int get_tasks_index_for_thread(em_proxying_queue* q, pthread_t thread) {
+  assert(q != NULL);
+  for (int i = 0; i < q->size; i++) {
+    if (pthread_equal(q->task_queues[i].thread, thread)) {
+      return i;
+    }
+  }
+  return -1;
+}
+
+// Not thread safe.
+static task_queue* get_or_add_tasks_for_thread(em_proxying_queue* q,
+                                               pthread_t thread) {
+  int tasks_index = get_tasks_index_for_thread(q, thread);
+  if (tasks_index != -1) {
+    return &q->task_queues[tasks_index];
+  }
+  // There were no tasks for the thread; initialize a new task_queue. If there
+  // are not enough queues, allocate more.
+  if (q->size == q->capacity) {
+    int new_capacity = q->capacity == 0 ? 1 : q->capacity * 2;
+    task_queue* new_task_queues =
+      realloc(q->task_queues, sizeof(task_queue) * new_capacity);
+    if (new_task_queues == NULL) {
+      return NULL;
+    }
+    q->task_queues = new_task_queues;
+    q->capacity = new_capacity;
+  }
+  // Initialize the next available task queue.
+  task_queue* tasks = &q->task_queues[q->size];
+  if (!task_queue_init(tasks, thread)) {
+    return NULL;
+  }
+  q->size++;
+  return tasks;
+}
+
+void emscripten_proxy_execute_queue(em_proxying_queue* q) {
+  assert(q != NULL);
+  pthread_mutex_lock(&q->mutex);
+  int tasks_index = get_tasks_index_for_thread(q, pthread_self());
+  task_queue* tasks = tasks_index == -1 ? NULL : &q->task_queues[tasks_index];
+  if (tasks == NULL || tasks->processing) {
+    // No tasks for this thread or they are already being processed.
+    pthread_mutex_unlock(&q->mutex);
```

Comment:
Given that tasks are dequeud before being run that mutex is unlocked.. why not allow for recursive running of the queue?    It certainly seems safe enough.

If we are going to fail due to recursive call perhaps we should return an error code so that we can detect that case when we have queued items but we are refusing the run them.

Reply:
The old proxying API prevents recursive queue processing, which makes sense given how frequently it is processed, although I don't whether it's necessary or not. Since we want to rewrite the old API in terms of the new one, I think it makes sense to be conservative and match the old behavior here, then see what happens if we try to relax it by allowing recursive processing in the future. Another option would be to make the system queue special in the new API and allow recursive processing of all queues except the system queue, but I would rather not special case the system queue like that.

Reply:
I'll add a TODO to investigate this, though.

Reply:
What about leaving out this logic and adding it instead to the old API when it calls into this one?   (I guess it could be as simple as a CAS with a `is_processing` global?).    Would make this API slightly simpler and lower level which seems in line with what you are going for maybe?

Reply:
Unfortunately this can't be an entirely separate layer. As part of the rewrite of the old API I would like to remove the postMessage handler for the old system and have everything go through the new handler. But the new handler would then need to call into a wrapper function for handling the recursion avoidance logic specific to the system queue. Since this recursion avoidance can't be entirely layered on top of the new system, it seems better to treat it uniformly across all queues. WDYT?

Reply:
I think it's possible that removing the recursion guard in a follow-up might Just Work, though. Here's the relevant comment from the old system:

>   // It is possible that when processing a queued call, the control flow leads back to calling this
  // function in a nested fashion! Therefore this scenario must explicitly be detected, and
  // processing the queue must be avoided if we are nesting, or otherwise the same queued calls
  // would be processed again and again.

You're right that the new system is different in that it dequeues before calling, so it won't run into this specific problem. I wouldn't be surprised if there are other reasons we currently depend on this behavior, so I would rather investigate as a follow-up.

Reply:
Yes, I even had a PR up for while to remove the nesting limitation.. but I agree there could be other semantic reasons why recursive running is not desirable.     I can't think of them.. but I suspect they exist.

lgtm to this change with or without the recursion restriction (perhaps mention in the docs that recursion doesn't currently work, and return failure when we fail to run for this reason).